engine/kv_engine_cleaner.cpp

```diff
@@ -235,366 +237,432 @@ void KVEngine::cleanNoHashIndexedSkiplist(
   }
 }
 
-struct PendingPurgeStrRecords {
-  PendingPurgeStrRecords(std::vector<StringRecord*>&& _records,
-                         TimeStampType _release_time)
-      : records(_records), release_time(_release_time) {}
-
-  std::vector<StringRecord*> records;
-  TimeStampType release_time;
-};
-
-struct PendingPurgeDLRecords {
-  PendingPurgeDLRecords(std::vector<DLRecord*>&& _records,
-                        TimeStampType _release_time)
-      : records(_records), release_time(_release_time) {}
-
-  std::vector<DLRecord*> records;
-  TimeStampType release_time;
-};
-
-void KVEngine::CleanOutDated(size_t start_slot_idx, size_t end_slot_idx) {
-  using ListPtr = std::unique_ptr<List>;
-  using HashListPtr = std::unique_ptr<HashList>;
-  TEST_SYNC_POINT_CALLBACK("KVEngine::backgroundCleaner::Start",
-                           &bg_work_signals_.terminating);
-  constexpr uint64_t kMaxCachedOldRecords = 1024;
-  constexpr size_t kSlotSegment = 1024;
-  constexpr double kWakeUpThreshold = 0.1;
-
-  std::deque<std::pair<TimeStampType, ListPtr>> outdated_lists;
-  std::deque<std::pair<TimeStampType, HashListPtr>> outdated_hash_lists;
-  std::deque<std::pair<TimeStampType, Skiplist*>> outdated_skip_lists;
-  std::deque<PendingPurgeStrRecords> pending_purge_strings;
-  std::deque<PendingPurgeDLRecords> pending_purge_dls;
+void KVEngine::prugeAndFreeAllType(
+    PendingPrugeFreeRecords& pending_clean_records) {
+  {  // purge and free pending string records
+    while (!pending_clean_records.pending_purge_strings.empty()) {
+      auto& pending_strings =
+          pending_clean_records.pending_purge_strings.front();
+      if (pending_strings.release_time <
+          version_controller_.LocalOldestSnapshotTS()) {
```

Comment:
What about GlobalOldestSnapshotTS()?

Reply:
Before commit (`ea3b1689184629f11ea617c512e5375e4ebcbc5e`), I split local snapshot (internal) and global snapshot (user hold).